src/shared/components/copyDownloadControls/CopyDownloadControls.tsx

```diff
@@ -0,0 +1,142 @@
+    });
+
+    return content.join('');
+}
+
+/**
+ * @author Selcuk Onur Sumer
+ * @author Aaron Lisman
+ */
+export class CopyDownloadControls extends React.Component<ICopyDownloadControlsProps, {}>
+{
+    private _copyButton: HTMLElement;
+
+    public static defaultProps:ICopyDownloadControlsProps = {
+        className: "",
+        showCopy: true,
+        showDownload: true,
+        autoGenerateHeader: false,
+        downloadFilename: "data.csv"
+    };
+
+    constructor()
+    {
+        super();
+        this.getText = this.getText.bind(this);
+        this.handleDownload = this.handleDownload.bind(this);
+    }
+
+    componentDidMount() {
+        // this is necessary because the clipboard wrapper library
+        // doesn't work with tooltips :(
+        if (this.props.showCopy  && this._copyButton) {
+            this.bindCopyButton();
+        }
+    }
+
+    public bindCopyButton()
+    {
+        new Clipboard(this._copyButton, {
+            text: function() {
+                return this.getText();
+            }.bind(this)
+        });
+    }
+
+    public render()
+    {
+        const arrowContent = <div className="rc-tooltip-arrow-inner"/>;
+
+        return (
+            <ButtonGroup className={this.props.className} style={{ marginLeft:10 }}>
+                <If condition={this.props.showCopy}>
+                    <DefaultTooltip
+                        overlay={<span>Copy</span>}
+                        placement="top"
+                        mouseLeaveDelay={0}
+                        mouseEnterDelay={0.5}
+                        arrowContent={arrowContent}
+                    >
+                        <button
+                            ref={(el) => {this._copyButton = el;}}
+                            className="btn btn-sm btn-default"
+                            data-clipboard-text="NA"
+                            id="copyButton"
+                            option-text={this.getText}
+                        >
+                            <i className='fa fa-clipboard'/>
+                        </button>
+                    </DefaultTooltip>
+                </If>
+
+                <If condition={this.props.showDownload}>
+                    <DefaultTooltip
+                        overlay={<span>Download CSV</span>}
+                        mouseLeaveDelay={0}
+                        mouseEnterDelay={0.5}
+                        placement="top"
+                        arrowContent={arrowContent}
+                    >
+                        <Button className="btn-sm" onClick={this.handleDownload}>
+                            <i className='fa fa-cloud-download'/>
+                        </Button>
+                    </DefaultTooltip>
+                </If>
+            </ButtonGroup>
+        );
+    }
+
+    public getText()
+    {
+        if (this.props.downloadDataGenerator) {
+            return serializeData(this.props.downloadDataGenerator() || [], this.props.autoGenerateHeader);
```

Comment:
would it make sense just to externalize serialization so that this isn't coupled with table data?  Have this component just accept a string or callback that takes a string?